File: leetcode/11-ContainerWithMostWater/solution.go
func maxArea(height []int) int {
    max := 0
    left, right := 0, len(height)-1
    for left < right {
        max = Max(max, (right - left) * Min(height[left], height[right]))
        if height[left] < height[right] {
            left += 1
        } else {
            right -= 1
        }
    }
    return max
}

func Max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

func Min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: leetcode/11-ContainerWithMostWater/solution.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut left= 0;
        let mut right= height.len()-1;

        let mut max = 0;

        while left < right {
            let width = (right - left) as i32;
            let lh = height[left];
            let rh = height[right];
            if lh <= rh {
                if max < lh * width {
                    max = lh * width;
                }
                left += 1;
            } else {
                if max < rh * width {
                    max = rh * width;
                }
                right -= 1;
            }
        }
        max
    }
}
